crates/autopilot/src/domain/competition/participation_guard/db.rs

```diff
@@ -0,0 +1,122 @@
+    /// timeout.
+    fn start_maintenance(
+        &self,
+        mut competition_updates_receiver: tokio::sync::mpsc::UnboundedReceiver<()>,
+        current_block: CurrentBlockWatcher,
+    ) {
+        let self_ = self.clone();
+        tokio::spawn(async move {
+            while competition_updates_receiver.recv().await.is_some() {
+                let current_block = current_block.borrow().number;
+                match self_
+                    .0
+                    .persistence
+                    .find_non_settling_solvers(self_.0.last_auctions_count, current_block)
+                    .await
+                {
+                    Ok(non_settling_solvers) => {
+                        let non_settling_drivers = non_settling_solvers
+                            .into_iter()
+                            .filter_map(|solver| {
+                                self_.0.drivers_by_address.get(&solver).map(|driver| {
+                                    Metrics::get()
+                                        .non_settling_solver
+                                        .with_label_values(&[&driver.name]);
+
+                                    driver.clone()
+                                })
+                            })
+                            .collect::<Vec<_>>();
+
+                        let non_settling_solver_names = non_settling_drivers
+                            .iter()
+                            .map(|driver| driver.name.clone())
+                            .collect::<Vec<_>>();
+
+                        tracing::debug!(solvers = ?non_settling_solver_names, "found non-settling solvers");
+
+                        let now = Instant::now();
+                        non_settling_drivers
+                            .into_iter()
+                            // Check if solver accepted this feature. This should be removed once a CIP is
+                            // approved.
+                            .filter_map(|driver| {
+                                driver.accepts_unsettled_blocking.then_some(driver.submission_address)
+                            })
+                            .for_each(|solver| {
+                                self_.0.banned_solvers.insert(solver, now);
+                            });
+                    }
+                    Err(err) => {
+                        tracing::warn!(?err, "error while searching for non-settling solvers")
+                    }
+                }
+            }
+        });
+    }
+}
+
+#[async_trait::async_trait]
+impl super::Validator for Validator {
+    async fn is_allowed(&self, solver: &eth::Address) -> anyhow::Result<bool> {
+        if let Some(entry) = self.0.banned_solvers.get(solver) {
+            if Instant::now().duration_since(*entry.value()) < self.0.ttl {
+                return Ok(false);
+            } else {
+                self.0.banned_solvers.remove(solver);
+            }
```

Comment:
I think this could lead to deadlocks. .get() keeps a read lock into the map while .remove() would require a write lock.
Given that the number of solvers that will ever enter the map is very small I think we might as well not remove the banned solver at all.
```suggestion
            return Ok(entry.elapsed() >= self.0.ttl);
```